### ISSUE TITLE
Allow app to sync data across devices

### ISSUE THREAD

**Issue by iSoron** [iSoron/uhabits#702]
## Description

Since version 2.0.0-alpha, Loop allows users to synchronize data across devices. This issue tracks a few remaining tasks that need to be completed before this feature can be released.

For general questions and comments about this feature, please post on #104. If you find bugs when using device sync, [please submit a new bug report](https://github.com/iSoron/uhabits/issues/new?assignees=&labels=&template=bug_report.md&title=). If you would like to help with the implementation, please get in touch in the [developer chat room](https://gitter.im/loophabits/dev).

## Development Tasks

- [x] **Implement and release initial prototype**
- [ ] **Server-side:**
	- [x] Generate sync links that expire after 15 minutes
	- [ ] Stop registrations if repository exceeds specified size
	- [ ] Enforce maximum upload size
	- [ ] Design explanation page for users who try to open a sync link in the browser
- [ ] **Client-side:**
	- [ ] Improve error handling:
		- [ ] Notify user that sync has been disabled due to unknown error
		- [ ] Automatically re-register key on `KeyNotFoundException`
		- [ ] Automatically re-upload data if decryption fails
	- [ ] Avoid loading entire database to memory
	- [ ] Switch from AES-CBC to AES-GCM
	- [ ] Switch to expiring sync links
	- [ ] Display last sync timestamp in settings
	- [ ] Handle habit deletion
- [ ] **Documentation:**
	- [ ] Write `docs/SYNC.md` explaining sync protocol
	- [ ] Update privacy policy
	- [ ] Update main app description

**Comment by necropolina** [comment]
@iSoron Would it be possible to add a "set the data directory" setting so users could set up their own sync capability via services like Syncthing?

Could be a quick way to enable sync for some users while a native sync feature is still being developed.

**Comment by dessalines** [comment]
2nd @necropolina 's suggestion. I looked in the `Android/data/org.isoron.uhabits` folder, and didn't find any files other than explicit backups.... where is the data currently stored?

**Comment by fairyex** [comment]
webdav might be a good solution.

**Comment by fairyex** [comment]
> 2nd @necropolina 's suggestion. I looked in the `Android/data/org.isoron.uhabits` folder, and didn't find any files other than explicit backups.... where is the data currently stored?

/data/data/,  You cannot ascess it without roots

**Comment by snoweuph** [comment]
> @iSoron webdav might be a good solution.

Yea, using webdav would also make it nextcloud compatible, would love that.

**Comment by SkypLabs** [comment]
In the GitHub Discussion related to this topic, I suggested the use of Etebase: https://github.com/iSoron/uhabits/discussions/104#discussioncomment-2221110.

**Comment by lkraav** [comment]
> > 2nd @necropolina 's suggestion. I looked in the `Android/data/org.isoron.uhabits` folder, and didn't find any files other than explicit backups.... where is the data currently stored?
> 
> /data/data/, You cannot ascess it without root

Also in https://github.com/iSoron/uhabits/discussions/104#discussioncomment-245999

> This implementation would require Loop to store the database in a public folder. Besides the security and privacy implications of this decision, it should also be noted that Android in general is [moving away from such practice](https://developer.android.com/preview/privacy/scoped-storage). It's not clear if this solution would still work after Android Q.

Do you think @iSoron you could be persuaded to accept a PR to make database storage location a user preference?

Even though I have a rooted phone (Sailfish OS ftw) and can access `/data/data/org.isoron.uhabits/databases/uhabits.db` with root credentials, it's a bit of jumping through hoops, and many cannot do this at all.

This would unlock people using the app to make progress on their goals and issues via syncing the database file wherever they want, while a more granular sync UX is under construction, with an unknown landing date (year even).

My use case: connect synced SQLite uhabits DB to an analytics tool like Metabase, to provide better daily parameter clarity to my health consultant to chase some health stuff. Isolating variables is a real challenge, and low-barrier tracking of inputs (habits) is something uHabits does really well :muscle:

**Comment by fillwithjoy1** [comment]
> > > 2nd @necropolina 's suggestion. I looked in the `Android/data/org.isoron.uhabits` folder, and didn't find any files other than explicit backups.... where is the data currently stored?
> > 
> > 
> > /data/data/, You cannot ascess it without root
> 
> Also in [#104 (comment)](https://github.com/iSoron/uhabits/discussions/104#discussioncomment-245999)
> 
> > This implementation would require Loop to store the database in a public folder. Besides the security and privacy implications of this decision, it should also be noted that Android in general is [moving away from such practice](https://developer.android.com/preview/privacy/scoped-storage). It's not clear if this solution would still work after Android Q.
> 
> Do you think @iSoron you could be persuaded to accept a PR to make database storage location a user preference?
> 
> Even though I have a rooted phone (Sailfish OS ftw) and can access `/data/data/org.isoron.uhabits/databases/uhabits.db` with root credentials, it's a bit of jumping through hoops, and many cannot do this at all.
> 
> This would unlock people using the app to make progress on their goals and issues via syncing the database file wherever they want via many mature 3rd party tools, while a more granular sync UX is under construction, with an unknown landing date (year even).
> 
> My use case: connect synced SQLite uhabits DB to an analytics tool like Metabase, to provide better daily parameter clarity to my health consultant to chase some health stuff. Isolating variables is a real challenge, and low-barrier tracking of inputs (habits) is something uHabits does really well 💪

Currently I'm using "Save a Copy" by Rikka to save it to my downloads folder, but yeah the exporting data is a lot of hoops to jump through

**Comment by jruiz94** [comment]
I agree with @lkraav , this syncing would be super easily implemented by just letting the users set a storage location so they can then use Syncthing for example to synchronize the habits.

This is pretty similar to what apps like Aegis do. In the particular case of Aegis, it lets you set a folder where automatic backups of your 2FA codes wil be stored on every change you make. You can then later set that folder to be synced with Syncthing.

**Comment by brog-io** [comment]
Hey, is there any update on this issue? It's very important to me since I love this app, but I'm afraid I'll lose my data

**Comment by Chickensoupwithrice** [comment]
If uHabits is already using sqlite, could something like [cr-sqlite](https://github.com/vlcn-io/cr-sqlite) be a solution here?

Optionally allowing users to host a websocket server to manage their own sync would be awesome. 

You could co conceivably charge money for a happy path for those who do not want to self host.

**Comment by Jefnull** [comment]
Are there any updates? I would happily pay for this feature. I love this app, but I need sync between my devices. I can't find any alternative apps that work quite the same. if there's anything we can do to accelerate this development, that would be great.

**Comment by technopagan** [comment]
Same request as others expressed above: syncing either with WebDav or Syncthing would be brilliant.
